### DISCUSSION TITLE
Regarding the splitting of the new method swanctl configuration file swanctl.conf

### DISCUSSION THREAD

**Discussion by yw1573** [strongswan/strongswan#849] [Q&A]
Dear maintainer：
  We used openswan a long time ago, and recently because of the new product project, we adopted strongswan-5.9.4 to support our new project with the latest swanctl.
  sorry my english is not very good.
  The following example：
```
|PC1[1.1.1.0/24]----[1.1.1.1/24]GWA[10.0.13.8/24]----[10.0.13.9/24]GWB[2.2.2.1/24]----[2.2.2.0/24]PC2
|                                      |
|                                      |
|                               [10.0.13.10/24]
|                                     GWC       
|                                [3.3.3.1/24]  
|                                      |             
|                                      |             
|                                [3.3.3.0/24]  
|                                     PC3         
```
  There will be two tunnels on GWA, GWA--GWB, GWA--GWC
  In the previous openswan, we used two configuration files, for example: [gwa_gwb.conf, gwa_gwb.secrets], [gwa_gwc.conf, gwa_gwc.secrets].
  gwa_gwb.conf:
```
conn gwa_gwb
    right=10.0.13.9
    rightsubnet=2.2.2.0/24
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    authby=secret
    esp=aes128-sha1;modp1024
    ike=aes128-sha1;modp1024
    compress=no
```
  gwa_gwb.secrets:
```
10.0.13.8 10.0.13.9 : PSK "www.strongswan.org"
```
  gwa_gwc.conf:
```
conn gwa_gwc
    right=10.0.13.10
    rightsubnet=3.3.3.0/24
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    authby=secret
    esp=aes128-sha1;modp1024
    ike=aes128-sha1;modp1024
    compress=no
```
  gwa_gwc.secrets:
```
10.0.13.8 10.0.13.10 : PSK "www.strongswan.org"
```
 I used ipsec stroke to build the tunnel, so I created gwa_gw_b.conf, gwa_gw_c.conf, gwa_gwb.scerts, gwa_gwc.scerts respectively which is almost the same as openswan
 However, this cannot successfully set up the GWA-GWB, GWA-GWC tunnel. If the GWA and GWB tunnels are successfully connected, GWC starts the connection with GWA, which will cause the GWA--GWB tunnel to fail to communicate. Among them, GWA and GWB do not send any interrupted tunnel information. Through the command on GWA, ipsec statusall can only see the tunnel information of GWC, but using ipsec statusall on GWB can see that the tunnel is normal.
  I don't know how to solve this problem. There is no such example on the official website, there is only one tunnel configuration file.
How can I modify the GWA configuration file to store GWA-GWB and GWA-GWC in different files, and the tunnel is established successfully

**Comment by tobiasbrunner** [comment]
> with the latest swanctl.

Then you don't want to configure ipsec.conf files but swanctl.conf (see [ConfigurationFiles](https://wiki.strongswan.org/projects/strongswan/wiki/ConfigurationFiles)). If you want to use separate files, the default swanctl.conf file includes *.conf files from the [`swanctl/conf.d`](https://wiki.strongswan.org/projects/strongswan/wiki/SwanctlDirectory) directory. For help on converting to swanctl.conf see [fromipsecconf](https://wiki.strongswan.org/projects/strongswan/wiki/fromipsecconf).

If you insist on using ipsec.conf, you need to include the config snippets in the default ipsec.conf file. Check that the configs are both loaded with `ipsec statusall`. Also, I don't see any `auto` settings in your snippets, did you set that in `conn %default`?

**Reply by yw1573** [reply]
I restarted my experimental project, the following is my configuration information, I use the stroke method first
```
# /etc/ipsec.conf - strongSwan IPsec configuration file
config setup

conn %default
    authby=secret
    keyexchange=ikev1
    ikelifetime=30m
    rekeymargin=3m
    keyingtries=1
conn 13_9
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    rigtht=10.0.13.9
    rightsubnet=2.2.2.0/24
    auto=add
conn 13_10
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    rigtht=10.0.13.10
    rightsubnet=3.3.3.0/24
    auto=add
# /etc/ipsec.secrets - strongSwan IPsec secrets file
10.0.13.8 10.0.13.9 : PSK "111"
10.0.13.8 10.0.13.10 : PSK "222"
```
When I make changes to ipsec.conf or ipsec.secrets, such as adding a tunnel when ipsec is running, I can successfully add the tunnel dynamically using ipsec rereadall.

**Reply by yw1573** [reply]
The above configuration is successful

**Reply by yw1573** [reply]
If I use stroke to split ipsec.conf, ipsec.secrets
```
# /etc/ipsec.conf - strongSwan IPsec configuration file
include /etc/ipsec/*.conf

# /etc/ipsec.secrets - strongSwan IPsec secrets file
include /etc/ipsec/*.secrets

# /etc/ipsec/default.conf - strongSwan IPsec configuration file
config setup

conn %default
    authby=secret
    keyexchange=ikev1
    ikelifetime=30m
    rekeymargin=3m
    keyingtries=1
# /etc/ipsec/13_9.conf - strongSwan IPsec configuration file
conn 13_9
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    right=10.0.13.9
    rightsubnet=2.2.2.0/24
    auto=add
# /etc/ipsec/13_10.conf - strongSwan IPsec configuration file
conn 13_10
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    right=10.0.13.10
    rightsubnet=3.3.3.0/24
    auto=add
    
# /etc/ipsec/13_9.secrets - strongSwan IPsec secrets file
10.0.13.8 10.0.13.9 : PSK "111"

# /etc/ipsec/13_10.secrets - strongSwan IPsec secrets file
10.0.13.8 10.0.13.10 : PSK "222"
```
I want to modify a configuration file under /etc/ipsec/ or I create a new tunnel configuration file, and I cannot load this information using ipsec rereadall

**Reply by yw1573** [reply]
That is to say, the modifications under /etc/ipsec/ cannot be captured by ipsec rereadall

**Reply by tobiasbrunner** [reply]
That's because that's not the command for it (see [ipsec](https://wiki.strongswan.org/projects/strongswan/wiki/ipseccommand)). Also, please don't use IKEv1 anymore.

**Reply by yw1573** [reply]
Yes, ipsec rereadall is used to reload certificates and keys, but how do I load the new tunnel configuration file under /etc/ipsec/ without affecting other tunnels, and I accept your suggestion, I Will fully use ikev2 in new projects

**Comment by yw1573** [comment]
Good morning @tobiasbrunner ：
 When I write in /etc/ipsec.conf and /etc/ipsec.secrets respectively:
```
# /etc/ipsec.conf - strongSwan IPsec configuration file
include /etc/ipsec/*.conf

# /etc/ipsec.secrets - strongSwan IPsec secrets file
include /etc/ipsec/*.secrets
```
 And write the tunnel configuration file connecting to GWB in the /etc/ipsec/ directory, and start IPSec. After the GWA-GWB tunnel is successfully established, I create a new file under /etc/ipsec/ and write The configuration file for connecting to GWC, how can I load the configuration file of the GWA-GWC tunnel and start the tunnel without affecting the GWA-GWB tunnel?

**Reply by tobiasbrunner** [reply]
I already linked the documentation for the `ipsec` command above (tldr: use `ipsec update`).

**Reply by yw1573** [reply]
I came up immediately after I received the email, but it was difficult for me to access github, and ipsec update could not load the tunnel configuration file I created under /etc/ipsec/

```
# /etc/ipsec.conf - strongSwan IPsec configuration file
include /etc/ipsec/*.conf

# /etc/ipsec.secrets - strongSwan IPsec secrets file
include /etc/ipsec/*.secrets

# /etc/ipsec/default.conf - strongSwan IPsec configuration file
conn %default
    authby=secret
    keyexchange=ikev1
    ikelifetime=30m
    rekeymargin=3m
    keyingtries=1

# /etc/ipsec/13_9.conf - strongSwan IPsec configuration file
conn 13_9
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    rigtht=10.0.13.9
    rightsubnet=2.2.2.0/24
    auto=add

# /etc/ipsec/13_9.secrets - strongSwan IPsec secrets file
10.0.13.8 10.0.13.9 : PSK "111"
```
I did command "ipsec restart"
Then I also executed the command "ipsec restart" on GWB and the tunnel of GWA-GWB was established successfully
Then the GWA-GWB tunnel is established successfully

Then I newly created 13_10.conf and 13_10.secrets under /etc/ipsec/
```
# /etc/ipsec/13_10.conf - strongSwan IPsec configuration file
conn 13_10
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    rigtht=10.0.13.10
    rightsubnet=3.3.3.0/24
    auto=add

# /etc/ipsec/13_10.secrets - strongSwan IPsec secrets file
10.0.13.8 10.0.13.10 : PSK "222"
```
Then I tried to load the tunnel configuration file of my newly created 13_10.conf using the command "ipsec update", but I failed, executing the command "ipsec statusall" did not see the configuration of 13_10 being loaded, and "ipsec up 13_10" also been warned

**Reply by tobiasbrunner** [reply]
There is definitely a typo there (`rigtht` should be `right` in both connections). And to read the secrets you also want to use `ipsec secrets` (or ìpsec rereadsecrets`). Other than that, I can't reproduce this. Make sure to read the logs.

**Reply by yw1573** [reply]
No, it's just a typo when I'm writing a review, there's really no problem with the configuration file, because our platform can't be copied, so I write the review by hand, all the configuration files are hand-typed, so it's inevitable There is an error, I want to check, but my network does not allow me, because it will cause me to fail to link to github

**Reply by yw1573** [reply]
I also saved the screenshot of the configuration file, and I am trying to find a way to link it

**Reply by yw1573** [reply]
I recompiled strongswan, and rearranged all the configuration files, forgive me, I should be negligent, after rearranging the configuration files, it is actually OK

**Reply by yw1573** [reply]
However, I found that when two tunnels are configured on one device with the same leftid and rightid settings, the two tunnels cannot exist at the same time
````
conn 13_9
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    right=10.0.13.9
    rightsubnet=2.2.2.0/24
    leftid=@jump1
    rightid=@jump2
    ike=aes128-sha1-modp1024!
    esp=aes128-sha1-modp1024!
    auto=add
conn 13_10
    left=10.0.13.8
    leftsubnet=1.1.1.0/24
    right=10.0.13.10
    rightsubnet=3.3.3.0/24
    leftid=@jump1
    rightid=@jump2
    ike=aes128-sha1-modp1024!
    esp=aes128-sha1-modp1024!
    auto=add
````
What should I do if I encounter such a problem？

**Reply by yw1573** [reply]
For details, see
https://github.com/yw1573/images/blob/master/strongswan/tunnel-collision/tunnel-collision.md

**Reply by tobiasbrunner** [reply]
Use different identities or disable `uniqueids` (see [ConfigSetupSection](https://wiki.strongswan.org/projects/strongswan/wiki/ConfigSetupSection)).

**Reply by yw1573** [reply]
Thank you so much
